### ISSUE TITLE
feat: More operators to strategy constraints

### ISSUE THREAD

**Issue by ivarconr** [Unleash/unleash#856]
> **This is an Unleash Enterprise feature.** 

Today strategy constraints supports `IN` and `NOT_IN` operators together with a list of values. There is a bunch of other operators that could make unleash even more flexible. 

Some examples:
- startsWidth - A string that starts with the given string
- endsWidth - A string that ends with the given string
- contains - A sting contains the string-fragment provided. 
- semver operators (app-version >= '1.2.2', etc)
- regex - We need to investigate a common subset of regex which can be uniformly modeled across programming languages (java, node, go, rust, python. js, etc)
- mathematical operators on numeric values (>, <, >=, <=) etc
- range operator

Which operator do you miss? 
https://docs.getunleash.io/docs/advanced/strategy_constraints

**Comment by PEZ** [comment]
Take this from someone who has not actually used Unleash, but anyway. 😄 

I wonder if it is really just an issue of adding operators? `IN/NOT_IN` are special because they gives us this option to list things. So for instance if I want to constraint on the domain-part of e-mail addresses, then `endsWidth` comes in handy. But what if I want to allow or disallow many domains?

Not that I know how to account for it, but maybe this specialness of `IN/NOT_IN` can be combined with the other operators? Something like:

```
endsWidth(IN('@foo.com', '@bar.org'))

endsWidth(NOT_IN('@blocked.com'))
```

I don't think it can be applied to all those operators, though.

The specific example could be solved by `regex` probably:

```
regex @(foo.com|bar.org)$
```

But that is more error prone than if it can be supported by `IN`.

Another alternative is to allow for pre-processing the context field, maybe using regex replace. Then the user could do:

```
match: ^(?:[^@]*@)
replace: ""
```

or

```
match: @(.*)$
replace: "$1"
```

And then the user could use any of the operators, inlcuding `IN/NOT_IN` on the result.

I think Google Analytics has something like this pre-processing option, but I could be remembering wrong.

**Comment by ivarconr** [comment]
We have decided to extend the operators Unleash supports, and aim for closing in on the specifications for the following operators. 

**List operators** (currently available)
- `IN`
- `NOT_IN`

**String Operators:**
- `STR_ENDS_WITH`
- `STR_STARTS_WITH`
- `STR_CONTAINS`

Notes: 
- string operators also supports a `caseInsensitive` field, which allow you to match on different cAsINg!

Examples:

```json
{
    "contextName": "customField",
    "operator": "STR_STARTS_WITH",
    "values": ["some-string"]

}

// Will be enabled when "customField" starts with "some-string"
```

```json
{
    "contextName": "email",
    "operator": "STR_ENDS_WITH",
    "values": ["@some-domain.com", "@another-domain.com"],
    "caseInsensitive": true
}

// Will be enabled when "email" field on the Unleash Context ends with "@some-domain.com" or "@another-domain.com"
```

Numeric operators:
- `NUM_EQ` - Value equal to the specified value. 
- `NUM_GT` - Value greater than the specified value. 
- `NUM_GTE` - Value equal to, or greater than, the specified value. 
- `NUM_LT `- Value lower than the specified value. 
- `NUM_LTE` - Value equal to, or lower than, the specified value. 

**Date operators:**
- `DATE_AFTER` - Any date after the specified time should be considered true. 
- `DATE_BEFORE` - Any date after the specified time should be considered true

Notes:
- Will assume json data format 
- Unleash will use an auto-generated `currentTime` on the Unleash Context. This value may be overridden by the SDK user. 

**Semver operators**
- SEMVER_EQ - Semver field equal to the specified value. E.g. 1.2.2
- SEMVER_GT - Semver field greater than the specified value. E.g. 1.2.2
- SEMVER_LT - Semver field lower than the specified value. E.g. 1.2.2

**General protocol enhancement*
- Some strategy constraints will use "value" property (not values). This applies to Numeric, Semver and data operators. 
- All constraints gets a new `inverted` property. This means we should invert the strategy constrain operator result. This makes it easy to say propertyA should not contain value "some", without us having to support a specific "STR_NOT_CONTAINS" operator. 

**Range support**
You can combining strategy constraints to create a "range" of legal values. This applies to all types.

Example:

```json
{
    "contextName": "currentTime",
    "operator": "DATE_AFTER",
    "value": "2022-01-22T13:00:00.000Z"
},
{
    "contextName": "currentTime",
    "operator": "DATE_BEFORE",
    "value": "2022-01-29T13:00:00.000Z"
}

// will be enabled when  `currentTime` is between the specified range
```
We need to implement these operators in all official SDKs. We have already established the specifications for the new feature in client-specifications ([PR](https://github.com/Unleash/client-specification/pull/25)) 🔥 

- [ ] Node.js SDK ([PR](https://github.com/Unleash/unleash-client-node/pull/289))
- [ ] Java SDK [PR](https://github.com/Unleash/unleash-client-java/pull/152)
- [ ] Unleash Proxy (comes via the  Node SDK)
- [ ] Python SDK
- [ ] .Net SDK
- [ ] Ruby SDK
- [ ] Go SDK
- [ ] Rust SDK
- [ ] PHP SDK

**Comment by chriswk** [comment]
This will also change the constraint object to
```js
{
  contextName: string,
  operator: string | operator,
  value: string,
  values: string[],
  inverse: bool,
  caseInsensitive: bool
}
```

**Comment by thomasheartman** [comment]
Nice! Did we settle on `inverse` for negating constraints? It's mentioned as `inverted` in the post above. Personally, I'm partial to `negate`, but as long as it's clearly documented, it doesn't matter too much, I think 💁🏼

**Comment by ivarconr** [comment]
It should be `inverted` @chriswk ! 

> Personally, I'm partial to negate

In general I agree, but the intention is to not require this field and also avoid the need to duplicate all the operators with the opposite. I also unsure if we will expose that option for all the operators. E.g. for NUM_LT you can just switch to NUM_GT if you want to negate it.

**Comment by chriswk** [comment]
Updated the object

**Comment by ivarconr** [comment]
released as part of v4.9

**Comment by rbtcollins** [comment]
Some implementation experience from the draft patch put up for the rust SDK: this model for the constraints is very hard to represent in the type system, which means that its fragile and hard to be confident about.

Secondly and separately, from experience at scale with Squid - rather than modelling a regex( which has very complicated θ behaviour depending on the combination of the supplied regex and the regex engine), to permit dealing with sets, consider explicit domain-aware set primitives: these can be optimised (e.g. via splay trees) to take advantage of task specific patterns (e.g. a tendancy for a small % of domains to be active at any one time, due to the usual power law of traffic popularity).

Happy to put either into new bugs if appropriate.
